sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala

```diff
@@ -1188,4 +1188,53 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
         classOf[BroadcastNestedLoopJoinExec]))
     }
   }
+
+  test("SPARK-32399: Full outer shuffled hash join") {
+    val inputDFs = Seq(
+      // Test unique join key
+      (spark.range(10).selectExpr("id as k1"),
+        spark.range(30).selectExpr("id as k2"),
+        $"k1" === $"k2"),
+      // Test non-unique join key
+      (spark.range(10).selectExpr("id % 5 as k1"),
+        spark.range(30).selectExpr("id % 5 as k2"),
+        $"k1" === $"k2"),
+      // Test string join key
+      (spark.range(10).selectExpr("cast(id * 3 as string) as k1"),
+        spark.range(30).selectExpr("cast(id as string) as k2"),
+        $"k1" === $"k2"),
+      // Test build side at right
+      (spark.range(30).selectExpr("cast(id / 3 as string) as k1"),
+        spark.range(10).selectExpr("cast(id as string) as k2"),
+        $"k1" === $"k2"),
+      // Test NULL join key
+      (spark.range(10).map(i => if (i % 2 == 0) i else null).selectExpr("value as k1"),
+        spark.range(30).map(i => if (i % 4 == 0) i else null).selectExpr("value as k2"),
+        $"k1" === $"k2"),
+      // Test multiple join keys
+      (spark.range(10).map(i => if (i % 2 == 0) i else null).selectExpr(
+        "value as k1", "cast(value % 5 as short) as k2", "cast(value * 3 as long) as k3"),
+        spark.range(30).map(i => if (i % 4 == 0) i else null).selectExpr(
+          "value as k4", "cast(value % 5 as short) as k5", "cast(value * 3 as long) as k6"),
+        $"k1" === $"k4" && $"k2" === $"k5" && $"k3" === $"k6")
+    )
+    inputDFs.foreach { case (df1, df2, joinExprs) =>
+      withSQLConf(
+        SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "80",
```

Comment:
You have used the word 'careful tuning' several times but I am not fully following out of my stupidity: Do you mean that to say that the broadcast threshold should be chosen to be larger than the build side size, so as to NOT trigger a BHJ ? 

Perhaps add a comment above this line to explain to the reader why the configs are chosen the way they are: For example why shuffle-partitions is 2 etc.

Btw, is it worth testing with shuffle partitions = 1 ?

Reply:
> I am wondering if you could add some "derivation" for the use of the magic constant 80 in the code comments ?

@agrawaldevesh - I feel it's hard. It depends on estimated size for the join child and we have multiple test cases here.

Reply:
Okay fine :-P the assert on line 1234 convinces me that the magic number "works". Tricking this thing to use SHJ is hard !.

Reply:
@agrawaldevesh - yes you feel my pain when crafting this kind of unit test :).